### ISSUE TITLE
How to unwind a #locsz of exactly 512 bytes

### ISSUE THREAD

**Issue by egelke** [MicrosoftDocs/cpp-docs#3830]
In the [Packed unwind data](https://docs.microsoft.com/en-us/cpp/build/arm64-exception-handling?view=msvc-160#packed-unwind-data) section, **point 5a** stipulates to use `stp x29,lr,[sp,#-locsz]!` for #locsz up and until 512.

_That is ok for the prolog, but how do you unwind that in the epilog?_

The obvious choice is off course to use `ldr x29,lr,[sp],#locsz` but that is an illegal assembly instruction since the index of `ldr` is limited to the range of [-512..504].  According to dumpbin the code should be `ldp  fp,lr,[sp],#0x200` with opcode #A8E07BFD but that is nonsense since opcode #A8E07BFD means a __negative__ #0x200 (I already created issue [#10012937](https://developercommunity.visualstudio.com/t/Packed-unwind-data-with-locsz--512-not/10012937) for that)

What should be done for this edge case?  Is it an oversight in the spec and should 5a be "CR == 11 && #locsz < 512" (_smaller then_ instead of smaller then or equal)?  If not, would it be possible to document the proper unwind instruction?  If it is an oversight, I assume it has a lot of implications on may produces including the part of the Windows OS that handles the packed unwind data and until that is fixed a #locsz of exactly 512 will never unwind correctly since epilog will be a different number of instructions then expected.

PS.  I didn't have the opportunity to test the actual unwinding yes; so far I only verified the executables with dumpbin.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 425e365b-d9a1-365e-d33f-0ab2645ce4f2
* Version Independent ID: 63f31be3-6a82-dd87-6d5d-9aa4e748703f
* Content: [ARM64 exception handling](https://docs.microsoft.com/en-us/cpp/build/arm64-exception-handling?view=msvc-160)
* Content Source: [docs/build/arm64-exception-handling.md](https://github.com/Microsoft/cpp-docs/blob/main/docs/build/arm64-exception-handling.md)
* Product: **visual-cpp**
* Technology: **cpp-tools**
* GitHub Login: @corob-msft
* Microsoft Alias: **corob**

**Comment by corob-msft** [comment]
@joemmett @pmsjt 
I'm not sure who to contact for clarification, so I'm reaching out to people who've done work on similar areas for comment.

**Comment by pmsjt** [comment]
It is unfortunate but there is a small window where `stp` can encore the displacement but  `ldp` cannot, as you correctly observed. To address this, the compiler needs to either over-allocate, so that prolog and epilog once again become symmetric, or produce an asymmetric prolog and epilog, and encode the function entry using the full formal (not the compact one).

For what is worth, MSVC prefers the later option. Here is an example:

```
      [RawPdata=000010E8 00002724]
   FuncLen=64  Vers=0  X=0  E=1  Epilogs=1  CodeWords=2
   Prolog unwind:
      01: BF...... +0000 stp   fp,lr,[sp,#-0x200]!; Actual=stp   fp,lr,[sp,#-0x200]!
      00: E1...... +0004 mov   fp,sp              ; Actual=mov   fp,sp
                   +0008 (end sequence)
   Epilog #1 unwind:  (Offset=auto  Index=3)
      03: 40...... +0058 ldp   fp,lr,[sp,#0x0]    ; Actual=ldp   fp,lr,[sp]
      04: C020.... +005C add   sp,sp,#0x200       ; Actual=add   sp,sp,#0x200
      06: E4...... +0060 end                      ; Actual=ret
                   +0068 (end sequence)
   [RawXdata=10E00019 40E4BFE1 E3E420C0]
```

**Comment by egelke** [comment]
@pmsjt, ok that makes sense.
@corob-msft, I think the above information would make a great addition to the documentation.

**Comment by pmsjt** [comment]
You are right @egelke . I am already planning on making some updates to the documentation as soon as I can find the time. For example, I need to add the new `save_any_reg` opcode for Arm64EC, and correct `save_next*` with is significantly incorrect. I'll try to add some clarifying lines to this case as well.
